In [2]:
import numpy as np
import pandas as pd
import scipy as sc
import re

## Day 10

In [122]:
input10 = pd.read_csv('day10_input', header=None).values.astype(int).flatten()
len(input10)

charging_outlet = 0
device_input = input10.max()+3

### Part A

In [172]:
sorted_adapters = np.sort(input10)
all_devices = [charging_outlet] + list(sorted_adapters) + [device_input]
all_devides = np.array(all_devides)
all_devides

array([  0,   1,   2,   3,   6,   9,  12,  13,  14,  15,  18,  19,  22,
        23,  26,  27,  28,  29,  32,  35,  36,  37,  40,  43,  46,  47,
        48,  49,  50,  53,  56,  57,  58,  59,  60,  63,  64,  65,  66,
        67,  70,  71,  72,  73,  74,  77,  78,  79,  80,  83,  86,  87,
        88,  89,  90,  93,  94,  95,  96,  97, 100, 103, 104, 105, 106,
       107, 110, 113, 114, 115, 118, 119, 122, 125, 128, 129, 130, 133,
       134, 135, 136, 139, 142, 143, 144, 145, 146, 149, 150, 151, 152,
       153, 156, 159, 162, 163, 166, 169, 170, 171, 172, 175, 178, 179,
       180, 183])

In [124]:
num_diff1 = (np.diff(all_devides)==1).sum()
num_diff2 = (np.diff(all_devides)==2).sum()
num_diff3 = (np.diff(all_devides)==3).sum()

In [125]:
answer10A = num_diff1*num_diff3
print(answer10A)

2574


### Part B

Find how many sub-sets of monotonically increasing numbers from the input set that add up to 183.

Find number of combinations of sets of consecutive numbers that differ by one. (at most 4 in this set)

Lengths of ones sets:
#### 1
Number of combinations: 1
#### 2
11 or 2

Number of combinations: 2
#### 3
111 21 12 3

Number of combinations: 4
##### 4
1111 112 121 211 22 31 13

Number of combinations: 7

In [167]:
import re

def number_of_combinations(length_ones):
    
    if length_ones < 2:
        return 1
    elif length_ones == 2:
        return 2
    elif length_ones == 3:
        return 4
    elif length_ones == 4:
        return 7
    else:
        raise ValueError

s=''.join(str(e) for e in np.diff(all_devices))
s_ones = re.split(r'3+',s)[:-1]
subs_ones_length = [len(s) for s in s_ones]
combinations_of_subsets = [number_of_combinations(nbr) for nbr in subs_ones_length]

In [171]:
answer10B = np.prod(combinations_of_subsets)
print(answer10B)

2644613988352
